disnake/ext/tasks/__init__.py

```diff
@@ -31,20 +31,47 @@
 import sys
 import traceback
 from collections.abc import Sequence
-from typing import Any, Callable, Coroutine, Generic, List, Optional, Type, TypeVar, Union
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Coroutine,
+    Generic,
+    List,
+    Optional,
+    Protocol,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+    get_origin,
+    overload,
+)
 
 import aiohttp
 
 import disnake
 from disnake.backoff import ExponentialBackoff
 from disnake.utils import MISSING, utcnow
 
+if TYPE_CHECKING:
+    from typing_extensions import Concatenate, ParamSpec, Self
+
+    P = ParamSpec("P")
+    CoroP = ParamSpec("CoroP")
+    Coro = Callable[CoroP, Coroutine[Any, Any, Any]]
```

Comment:
This should strip the self type from `CoroP`, otherwise type-checkers will expect you to use `self.background_task(self, x, y)` in classes (which doesn't work) instead of `self.background_task(x, y)`.

This also came up [here](https://github.com/DisnakeDev/disnake/pull/675#discussion_r968657205) a few days ago, however that one's a bit more clear as it only needs to account for `self` being a `Cog` instance, while `Loop` could technically be used without cogs entirely.
I don't think there's a way of reliably stripping the self type from a paramspec regardless of the type, so maybe it's sufficient for now to only account for cogs as well.

tl;dr, something along the lines of this would be needed
```py
    InjectionCallback = Union[
        Callable[Concatenate[CogT, P], T_],
        Callable[P, T_],
    ]
```

Reply:
I'm looking at the examples, but I don't think it can be done in a manner that'd work with inheritance; the `__get__` would need to look something like
```py
@overload
def __get__(self, instance: None, owner: Any) -> Loop[CoroP]:
    ...
@overload
def __get__(self, instance: T, owner, Any) -> Loop[Concatenate[T, CoroP]]:
    ....
```
...which in both definitions hardcodes the return obj type as `Loop`
*There \*might\* be some sneaky way using Protocols (my beloved) to hint the return type as `SomeProto[Self, CoroP]` but I don't know how it is going to work here*

Reply:
Yeahh exactly, that'd work if `LoopT[CoroP]` was a thing or if one could rebind `Self`, which unfortunately isn't the case.
The [PEP](https://peps.python.org/pep-0673/) even explicitly states this:
> Note that we reject using Self with type arguments, such as Self[int].
> ...
> In such cases, we recommend using an explicit type for self

... which doesn't work, since we *don't* want an explicit type for self here :/

I've been trying to find an alternative solution, spent the last hour trying different things and looking through mypy/pyright bugtrackers and stackoverflow, but it seems this just isn't possible without HKT.

For now it's probably best to go back to `Loop[LF]`, and type all the `*args` and `**kwargs` as `Any`.

Reply:
Done as suggested

Reply:
https://github.com/DisnakeDev/disnake/pull/655/commits/2215240044a4fdd379a891b2e1681fa5525c7caf